client/src/erlang/receiver.zig

```diff
@@ -0,0 +1,311 @@
+    var length: i32 = undefined;
+    var ty: i32 = undefined;
+    try erl.validate(error.decoding_atom_string_length, ei.ei_get_type(deserializer.buf.buff, deserializer.index, &ty, &length));
+
+    if (ty != ei.ERL_STRING_EXT and ty != ei.ERL_ATOM_EXT)
+        return error.message_is_not_atom_or_string;
+
+    const u_length: u32 = @bitCast(length);
+
+    const buffer = try allocator.allocSentinel(u8, u_length, 0);
+    try erl.validate(error.decoding_atom, erlang_fun(deserializer.buf.buff, deserializer.index, buffer.ptr));
+    return buffer;
+}
+
+fn receive_string(deserializer: Deserializer, allocator: std.mem.Allocator) ![:0]const u8 {
+    return receive_atom_string(deserializer, allocator, ei.ei_decode_string);
+}
+
+fn receive_atom(deserializer: Deserializer, allocator: std.mem.Allocator) ![:0]const u8 {
+    return receive_atom_string(deserializer, allocator, ei.ei_decode_atom);
+}
+
+pub fn run_with_pid(comptime T: type) type {
+    return std.meta.Tuple(&.{ ei.erlang_pid, T });
+}
+
+pub fn run(comptime T: type, allocator: std.mem.Allocator, ec: *erl.Node) !T {
+    var msg: ei.erlang_msg = undefined;
+    var buf: ei.ei_x_buff = undefined;
+    var index: i32 = 0;
+
+    try erl.validate(error.create_new_decode_buff, ei.ei_x_new(&buf));
+
+    while (true) {
+        const got: i32 = ei.ei_xreceive_msg(ec.fd, &msg, &buf);
+        if (got == ei.ERL_TICK)
+            continue;
+        if (got == ei.ERL_ERROR) {
+            return error.got_error_receiving_message;
+        }
+        break;
+    }
+
+    try erl.validate(error.decoding_version, ei.ei_decode_version(buf.buff, &index, null));
+    return internal_receive_message(T, allocator, .{
+        .buf = &buf,
+        .index = &index,
+        .ec = ec,
+    });
+}
+
+inline fn receive_struct(comptime T: type, comptime item: std.builtin.Type.Struct, allocator: std.mem.Allocator, deserializer: Deserializer) !T {
+    var value: T = undefined;
+    var size: i32 = 0;
+    if (item.is_tuple) {
+        try erl.validate(
+            error.decoding_tuple,
+            ei.ei_decode_tuple_header(deserializer.buf.buff, deserializer.index, &size),
+        );
+        if (item.fields.len != size) return error.wrong_tuple_size;
+        inline for (&value) |*elem| {
+            elem.* = try internal_receive_message(@TypeOf(elem.*), allocator, deserializer);
+        }
+    } else {
+        try erl.validate(
+            error.decoding_map,
+            ei.ei_decode_map_header(deserializer.buf.buff, deserializer.index, &size),
+        );
+        const fields = std.meta.fields(T);
+        var present_fields: [fields.len]bool = .{false} ** fields.len;
+        var counter: u32 = 0;
+        if (size > fields.len) return error.too_many_map_entries;
+        for (0..@as(u32, @bitCast(size))) |_| {
+            const key = try receive_atom(deserializer, allocator);
+            inline for (0.., fields) |idx, field| {
```

Comment:
There's probably a way to avoid this loop.

Reply:
Added as a TODO.